### DISCUSSION TITLE
How to convert places.geometry to PostGis geometry

### DISCUSSION THREAD

**Discussion by lilcorey** [OvertureMaps/data#69] [Discussion]
I copied `places` data into Postgres. `geometry` is stored as text (some samples below). I'm having trouble converting `geometry` to a valid Geometry object with PostGIS. The [example](https://postgis.net/docs/ST_GeomFromWKB.html) on PostGIS docs show `geometry`'s binary string representation differently.
```
ST_GeomFromWKB(E'\\001\\002\\000\\000\\000\\002\\000\\000\\000\\037\\205\\353Q\\270~\\\\\\300\\323Mb\\020X\\231C@\\020X9\\264\\310~\\\\\\300)\\\\\\217\\302\\365\\230C@',4326)
```
How can I work with string representation of places data in PostGIS?

**Sample geometry rows from Postgres**
```
 \x01\x01\x00\x00\x00\x97\xAE`\x1B\xF1iU\xC0&\x1D\xE5`6WF@
 \x01\x01\x00\x00\x00\xA9\x9F7\x15\xA9\x14T\xC0\xC1\xAD\xBBy\xAAS;@
 \x01\x01\x00\x00\x00\xB9\xA8\x16\x11E\x9BR\xC0\x1C6\xDB&&dD@
```

**Queries tried**
```
postgres=# select ST_GeomFromWKB(decode(geometry, 'escape')) from places;
ERROR:  invalid input syntax for type bytea

postgres=# select ST_GeomFromWKB(decode(geometry, 'hex')) from places;
ERROR:  invalid hexadecimal digit: "\"

postgres=# select ST_GeomFromWKB(decode(replace(geometry, '\', '\\'), 'escape')) from places;
ERROR:  Invalid endian flag value encountered.

postgres=# select ST_GeomFromWKB(decode(replace(geometry, '\', '\\'), 'hex')) from places;
ERROR:  invalid hexadecimal digit: "\"
```

**(edit) ANSWER**
Use the `spatial` DuckDB package.
```
INSTALL spatial;
LOAD spatial;
SELECT  ST_GeomFromWkb(geometry) AS geometry FROM ...
```

this [resource](https://til.simonwillison.net/overture-maps/overture-maps-parquet) was helpful for using `spatial`

**Comment by songololo** [comment]
I believe `postGIS` expects a hex representation of WKB geometry. If you're working with Python you can use `shapely` to extract the `wkb` representation. Going by memory but you would either need to look for a `wkb_hex` property or possibly need to set a `hex=True` parameter.

**Comment by mtravis** [comment]
@lilcorey How are loading the parquet into postigis? I have method using ogr2ogr that I can share if that would be useful?

**Reply by lilcorey** [reply]
My full flow is as follows
1) Using DuckDB, copy all `places` to CSV (do light formatting on certain cols that cause problems in the next step e.g. escape chars)
2) Partition single `places` CSV into smaller files
2) Use postgres CLI to copy CSVs to Postgres

If you can share I'll take a look. I haven't found a solution yet but I think if I go back and use the `spatial` import in DuckDB I can get the Geometry. I may still have the issue of casting string to the right type in Postgres though. Will update when I can

**Comment by mtravis** [comment]
@lilcorey aplogies for never getting back to you on this did you ever get this working? 

My workflow has changed a little due to the data changing but I normally export to parquet using DuckDB and then run this command in `ogr2ogr`

```bash
ogr2ogr -overwrite -f postgresql PG:"dbname=staging" admins.parquet -nln acadmn.raw_admin \
 -sql "SELECT id,name,updateTime,version,adminLevel,subType,localityType,,isoCountryCodeAlpha2,isoSubCountryCode,defaultLanguage,sources,ST_GeomFromWKB(geometry,4326)  as geom from admins" \
 -dialect sqlite --config PG_USE_COPY YES -progress 
```

**Reply by lilcorey** [reply]
NP!

I initially used CSV w/ binary format because I didn't want to use any additional imports in DuckDB but have opted for geoJSON export format and have successfully exported geometry with `spatial` import. I'm now investigating `org2ogr` to get the geoJSON export into postgres.

If your `ogr2ogr` input data is parquet I'm curious how you're using DuckDB? Is it needed? Is it easy as pointing `ogr2ogr` to the OG overture parquet data to import to postgres?! *edit:* I'm on gdal-bin version 3.2.2 and it seems parquet support starts in v3.6. I'm doing this in a postgis docker container and so far only have v3.2.2

**Reply by mtravis** [reply]
@lilcorey  I use duckdb to cut the area of data, e.g. bbox of UK, and then export that to a parquet file using duckdb and then load to postgis using ogr2ogr like above.

gdal version is an issue. I have installed a newer version (3.7) using Conda like so
```
conda install -c conda-forge gdal libgdal-arrow-parquet
``` 

I guess you might be able to add docker to your container somehow?

**Comment by lilcorey** [comment]
### GeoJSON
First, load duckdb. We're going to export the data to GeoJSON. I recommend filtering the dataset otherwise the file size will explode. Parquet compresses data really well.
```
INSTALL spatial;
LOAD spatial;
COPY ( SELECT
  id,
  updatetime,
  version,
  JSON(names) as names, 
  JSON(categories) as categories, 
  confidence, 
  JSON(websites) as websites, 
  JSON(socials) as socials,
  JSON(emails) as emails,
  JSON(phones) as phones, 
  JSON(brand) as brand, 
  JSON(addresses) as addresses, 
  JSON(sources) as sources,
  ST_GeomFromWKB(geometry) as geometry 
FROM read_parquet('overture/places/*', filename=true)) 
TO 'places.geojson' WITH (FORMAT GDAL, DRIVER 'GeoJSON');
```

This will give you an exported `places.geojson` file.
Next, use `ogr2ogr` to convert GeoJSON to Postgres. At this point a db instance should exist already. You need to know the db name and username. This example below assumes a table doesn't exist. `-nln places` creates a table named places. `ogr2ogr` assumes a schema based on the input data.

```
ogr2ogr -f "PostgreSQL" PG:"dbname=postgres user=postgres" "var/lib/postgresql/data/places.geojson" -nln places
```
### Parquet
As suggested by @mtravis it seems `ogr2ogr` has direct parquet support as of v3.5 or v3.6. I haven't been able to test this method but adding it here as an accepted answer since it is the more direct method for converting places Parquet to Postgres (postgis). See their comment above for an example.
